### PR TITLE
Proof of concept Pre-Authentication flow

### PR DESCRIPTION
**Pull request by sgmiller**
Adds a special error return, DelegatedAuthenticationError, which can be used to signal to the request processing system that an internal call to an auth mount is needed before re-processing this request.  In this spike the mount accessor is provided by the backend, and so would have to be a configuration element. 

In addition, I added a mount tuneable to allow access to mounts for pre-auth purposes.  This may or may not be overkill.

When encountering the error, the pre-auth mount is resolved, and any extra data the backend likes is tacked onto the auth request.  The auth method is called, and if successful, the client token is re-set on the initial request and the initial request is re-handled. 

API endpoints that need pre-auth have to be declared Unauthenticated, and detect that they need pre-auth, returning the above error, and then subsequently detecting that auth occurred (and occurred internally), to finish processing the call.

### CODE REVIEW

File: vault/mount.go [hunk 1]
```diff
@@ -294,6 +295,21 @@ func (t *MountTable) findByBackendUUID(ctx context.Context, backendUUID string)
...
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < n; i++ {
+		if entry := t.Entries[i]; entry.Accessor == accessor && entry.Namespace().ID == ns.ID {
+			return entry, nil
+		}
+	}
```
Comment (sgmiller) [inline vault/mount.go:309]: Might want to make a map/cache of this in case there are a lot of auth mounts.

File: vault/request_handling.go [hunk 2]
```diff
@@ -950,7 +952,7 @@ func (c *Core) doRouting(ctx context.Context, req *logical.Request) (*logical.Re
 }
 
 func (c *Core) isLoginRequest(ctx context.Context, req *logical.Request) bool {
-	return c.router.LoginPath(ctx, req.Path)
+	return c.router.LoginPath(ctx, req.Path) && req.ClientToken == ""
```
Comment (sgmiller) [inline vault/request_handling.go:955]: This change seemed necessary but makes me nervous.

File: vault/mount.go (hunk 1 above)
Comment (stevendpclark) [reply vault/mount.go:309]: Kinda surprised we don't need a read lock for the iteration across t.Entries to prevent data races

Comment (sgmiller) [reply vault/mount.go:309]: I believe this works because MountTables are replaced wholesale every time they're changed, they're not modified in place.

**Pull request closed without merging**